docs/updates/5-1-2017-flow.md

```diff
@@ -0,0 +1,156 @@
+### Flow Coverage Report (May 1st)
+
+`yarn flow-coverage`
+
+Stats:
+* Flow version `0.43.1`
+* **60 %**
+
+#### Components
+
+* **54%** coverage
+* 44 components are typed (almost all)
+* `yarn flow-react` - creates a flow report for components
+
+Typing our components helps us with rendering logic, which can take advantage
+of our domain objects such as `Frame`.
+
+```js
+function renderFrameTitle(frame: Frame) {
+  const displayName = formatDisplayName(frame);
+  return dom.div({ className: "title" }, displayName);
+}
+```
+
+We are not yet typing our props and state, which would be a helpful way
+to get additional coverage. We currently use `PropTypes`, which do not include
+flow types. We will hopefully switch to [props] soon to take advantage of static typing.
+
+```js
+shouldComponentUpdate(nextProps, nextState) {
+  const { frames, selectedFrame } = this.props;
+  const { showAllFrames } = this.state;
+  return (
+    frames !== nextProps.frames ||
+    selectedFrame !== nextProps.selectedFrame ||
+    showAllFrames !== nextState.showAllFrames
+  );
+}
+```
+
+There have been a couple places where flow types have been difficult to get right.
+
+1. **contravariants** `this.toggleFramesDisplay = this.toggleFramesDisplay.bind(this)`
+2. **defaultProps** and other class properties.
+
+Overall, typing our components has been a nice win though as it gives us additional confidence our UI will behave appropriately and catches several lifecycle edge-cases
+where properties could be null or racey.
+
+#### Utilities
+
+* **59%** coverage
+* **60** files
+* `yarn flow-utils` - creates a flow report for utils
+
+We started typing our utility functions. Starting with utilities helped us
+document our primary business logic and made them easier to refactor.
+
+In this example, we added types to a parser utility that checks to see if an
+expression is in a particular scope.
+
+```js
+export function isExpressionInScope(expression: string, scope?: Scope) {
+  if (!scope) {
+    return false;
+  }
+
+  const variables = getVariablesInScope(scope);
+  const firstPart = expression.split(/\./)[0];
+  return variables.includes(firstPart);
+}
+```
+
+We are not currently typing our test files, but I think it would be helpful if
+we did. In this example, we are searching for all of the functions in a given
+source, but because we did not provide a type for `getSourceText`, flow can
+not use `getSymbols`'s types to tell us if we're writing a reasonable test.
```

Comment:
Generally type systems eliminate the need to do things like that. But you can do one of two things in that scenario:

1. Use Flow's [`suppress_comment` option](https://flow.org/en/docs/config/options/#toc-suppress-comment-regex)
2. Pass the value through `any` to stop type checking – I'd recommend doing this along with a comment explaining why it `any` is needed. Using `any` means you're less tied to configuration

```js
function method(value: string) {
  if (typeof value === 'number') throw new TypeError('Support for numbers has been removed');
  return value;
}

test('throws deprecation error', () => {
  expect(() => {
    // Passing through `any` because we want to make sure we're throwing an error.
    method((42: any));
  }).toThrow(TypeError);
});
```